src/client/dfs/dcache.c

```diff
@@ -0,0 +1,992 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#define D_LOGFAC DD_FAC(dfs)
+
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <inttypes.h>
+
+#include <daos/debug.h>
+#include <gurt/hash.h>
+#include <gurt/list.h>
+#include <daos/common.h>
+
+#include "dfs_internal.h"
+
+#define DF_TS                "%ld.%09ld"
+#define DP_TS(t)             (t).tv_sec, (t).tv_nsec
+
+#ifdef DAOS_BUILD_RELEASE
+
+#define DF_DK      "dk[%zi]"
+#define DP_DK(_dk) strnlen((_dk), DCACHE_KEY_MAX)
+
+#else
+
+#define DF_DK      "dk'%s'"
+#define DP_DK(_dk) daos_dk2str(_dk)
+
+static inline char *
+daos_dk2str(const char *dk)
+{
+	int         i, j;
+	const char *it;
+	const char *sep = "-:";
+
+	if (dk == NULL)
+		return "<NULL>";
+
+	it = dk;
+	for (i = 0; i < 2; i++) {
+		for (j = 0; j < 16; j++) {
+			if (it[j] < '0' || it[j] > 'f')
+				return "<invalid dentry key>";
+		}
+		it += 16;
+		if (it[0] != sep[i])
+			return "<invalid dentry key>";
+		it++;
+	}
+
+	for (i = 0; i < NAME_MAX; i++) {
+		if (it[i] == '\0')
+			return (char *)dk;
+		if (!isprint(it[i]) || it[i] == '\'')
+			return "<not printable>";
+	}
+	return "<dentry key too long>";
+}
+
+#endif
+
+typedef int (*destroy_fn_t)(dfs_dcache_t *);
+typedef int (*find_insert_fn_t)(dfs_dcache_t *, char *, size_t, int, dfs_obj_t **, mode_t *,
+				struct stat *);
+typedef int (*find_insert_rel_fn_t)(dfs_dcache_t *, dfs_obj_t *, const char *, size_t, int,
+				    dfs_obj_t **, mode_t *, struct stat *);
+typedef void (*drec_incref_fn_t)(dfs_dcache_t *, dfs_obj_t *);
+typedef void (*drec_decref_fn_t)(dfs_dcache_t *, dfs_obj_t *);
+typedef void (*drec_del_at_fn_t)(dfs_dcache_t *, dfs_obj_t *);
+typedef int (*drec_del_fn_t)(dfs_dcache_t *, char *, dfs_obj_t *);
+
+/** DFS directory cache */
+struct dfs_dcache {
+	/** Cached DAOS file system */
+	dfs_t              *dd_dfs;
+	/** Hash table holding the cached directories */
+	struct d_hash_table dd_dir_hash;
+	/** Key prefix of the DFS root directory */
+	char                dd_key_root_prefix[DCACHE_KEY_PREF_SIZE];
+	/** flag to indicate whether garbage collection is disabled or not */
+	bool                 dd_disable_gc;
+	/** Garbage collector time-out of a dir-cache record in seconds */
+	uint32_t            dd_timeout_rec;
+	/** Entry head of the garbage collector list */
+	d_list_t             dd_head_gc;
+	/** Mutex protecting access to the garbage collector list */
+	pthread_mutex_t      dd_mutex_gc;
+	/** Size of the garbage collector list */
+	uint64_t             dd_count_gc;
+	/** Time period of garbage collection */
+	uint32_t             dd_period_gc;
+	/** Maximal number of dir-cache record to reclaim per garbage collector trigger */
+	uint32_t             dd_reclaim_max_gc;
+	/** Next Garbage collection date */
+	struct timespec      dd_expire_gc;
+	/** True iff one thread is running the garbage collection */
+	atomic_flag          dd_running_gc;
```

Comment:
In pil4dfs dcache, the type of this field and its initialization is invalid.
If you want, I can fix it or do you want to do it.

Reply:
i had to actually change how this is done because it was complaining when compiling with clang. im not sure why it did not complain in pil4dfs though.
anyway it would be great if you fix the pil4dfs one in master, and i can bring it over when i merge latest master.

Reply:
Me to do not understand why it did not complain with the pil4dfs patch.
I have indeed reuse this kind of stuff in a latter PR and have the same error message with clang.

Reply:
Draft of pull request created #14995.
Waiting for the result of the CI before submitting if to official review.